File: src/main/java/net/igneo/icv/enchantmentActions/PlayerEnchantmentActionsProvider.java
package net.igneo.icv.enchantmentActions;

import net.minecraft.core.Direction;
import net.minecraft.nbt.CompoundTag;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityManager;
import net.minecraftforge.common.capabilities.CapabilityToken;
import net.minecraftforge.common.capabilities.ICapabilityProvider;
import net.minecraftforge.common.util.INBTSerializable;
import net.minecraftforge.common.util.LazyOptional;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class PlayerEnchantmentActionsProvider implements ICapabilityProvider, INBTSerializable<CompoundTag> {

    public static Capability<PlayerEnchantmentActions> PLAYER_ENCHANTMENT_ACTIONS = CapabilityManager.get(new CapabilityToken<PlayerEnchantmentActions>() {});

    private PlayerEnchantmentActions enchNBT = null;
    private final LazyOptional<PlayerEnchantmentActions> optional = LazyOptional.of(this::createEnchantmentActions);

    private PlayerEnchantmentActions createEnchantmentActions() {
        if (this.enchNBT == null) {
            this.enchNBT = new PlayerEnchantmentActions();
        }
        return this.enchNBT;
    }

    @Override
    public @NotNull <T> LazyOptional<T> getCapability(@NotNull Capability<T> cap, @Nullable Direction side) {
        if (cap == PLAYER_ENCHANTMENT_ACTIONS) {
            return optional.cast();
        }
        return LazyOptional.empty();
    }

    @Override
    public CompoundTag serializeNBT() {
        CompoundTag nbt = new CompoundTag();
        createEnchantmentActions().saveNBTData(nbt);
        return nbt;
    }

    @Override
    public void deserializeNBT(CompoundTag nbt) {
        createEnchantmentActions().loadNBTData(nbt);
    }
}

File: src/main/java/net/igneo/icv/enchantmentActions/PlayerEnchantmentActionsProvider.kt
package net.igneo.icv.enchantmentActions

import net.minecraft.core.Direction
import net.minecraft.nbt.CompoundTag
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.CapabilityManager
import net.minecraftforge.common.capabilities.CapabilityToken
import net.minecraftforge.common.capabilities.ICapabilityProvider
import net.minecraftforge.common.util.INBTSerializable
import net.minecraftforge.common.util.LazyOptional

class PlayerEnchantmentActionsProvider : ICapabilityProvider, INBTSerializable<CompoundTag> {
    private var enchNBT: PlayerEnchantmentActions? = null
    private val optional = LazyOptional.of { createEnchantmentActions() }

    private fun createEnchantmentActions() = enchNBT ?: PlayerEnchantmentActions().also { enchNBT = it }

    override fun <T : Any?> getCapability(cap: Capability<T>, side: Direction?): LazyOptional<T> =
        if (cap === PLAYER_ENCHANTMENT_ACTIONS) optional.cast() else LazyOptional.empty()

    override fun serializeNBT(): CompoundTag = CompoundTag().also {
        createEnchantmentActions().saveNBTData(it)
    }

    override fun deserializeNBT(nbt: CompoundTag) {
        createEnchantmentActions().loadNBTData(nbt)
    }

    companion object {
        val PLAYER_ENCHANTMENT_ACTIONS: Capability<PlayerEnchantmentActions> =
            CapabilityManager.get(object : CapabilityToken<PlayerEnchantmentActions>() {})
    }
}
